sympy/solvers/ode.py

```diff
@@ -4089,6 +4089,25 @@ def ode_nth_order_reducible(eq, func, order, match):
 
     return fsol
 
+# This needs to produce an invertible function but the inverse depends
+# which variable we are integrating with respect to. Since the class can
+# be stored in cached results we need to ensure that we always get the
+# same class back for each particular integration variable.
+def _nth_algebraic_diffx(var, _stored={}):
+    cls = _stored.get(var, None)
+
+    if cls is None:
+        # A class that behaves like Derivative wrt var but is "invertible".
+        class diffx(Function):
+            def inverse(self):
+                # don't use integrate here because fx has been replaced by _t
+                # in the equation; integrals will not be correct while solve
+                # is at work.
+                return lambda expr: Integral(expr, var) + Dummy('C')
+
+        _stored[var] = cls = diffx
```

Comment:
I'm wondering now whether this is threadsafe. I think maybe the above should be:
```
cls = _stored.setdefault(var, diffx)
```

Reply:
I'm not sure this would be completely safe, either. `__hash__` and `__eq__` have to be called for `var`. (cf, https://google.github.io/styleguide/pyguide.html#218-threading) But still, this looks better.

Reply:
Nothing more to add from me. I'll leave this for you guys to commit.